cmd/host/src/lib.rs

```diff
@@ -249,103 +219,62 @@ fn host_last_panic(hubris: &HubrisArchive, core: &mut dyn Core) -> Result<()> {
-            0xeba9 => "IPCC_PANIC_EARLYBOOT_TRAP".to_owned(),
-            0xebff => "IPCC_PANIC_EARLYBOOT_*".to_owned(),
-            b => format!("Unknown `ipd_cause` {b:#06x}"),
-        }
-    );
-
-    let ipd_error = p.ipd_error;
-    let ipd_cpuid = p.ipd_cpuid;
-    let ipd_thread = p.ipd_thread;
-    let ipd_addr = p.ipd_addr;
-    let ipd_pc = p.ipd_pc;
-    let ipd_fp = p.ipd_fp;
-    let ipd_rp = p.ipd_rp;
-
-    println!("ipd_error:   {ipd_error}");
-    println!("ipd_cpuid:   {ipd_cpuid}");
-    println!("ipd_thread:  {ipd_thread:#x}");
-    println!("ipd_addr:    {ipd_addr:#x}");
-    println!("ipd_pc:      {ipd_pc:#x}");
-    println!("ipd_fp:      {ipd_fp:#x}");
-    println!("ipd_rp:      {ipd_rp:#x}");
-
-    match std::str::from_utf8(&p.ipd_message) {
-        Ok(s) => println!("ipd_message: {}", s.trim_matches('\0')),
-        Err(e) => println!(
-            "ipd_message: {:?}\n  (could not decode: {e})",
-            p.ipd_message
-        ),
-    }
-    println!("ipd_stackid: {}", p.ipd_stackidx);
-    println!("stack trace:");
-    let syms: Vec<String> = p
-        .ipd_stack
-        .iter()
-        .map(|s| {
-            let sym = match std::str::from_utf8(&s.ips_symbol) {
-                Ok(s) => s.to_owned(),
-                Err(e) => format!("<could not decode {:?}: {e}>", s.ips_symbol),
-            };
-            let offset = s.ips_offset;
-            format!("{}+{offset:#x}", sym.trim_matches('\0'))
-        })
-        .collect();
-    let width = syms.iter().map(|s| s.len()).max().unwrap_or(0);
-    for (sym, s) in syms.iter().zip(p.ipd_stack.iter()) {
-        let addr = s.ips_addr;
-        println!("  {sym:width$}  ({addr:#016x})");
+
+    if let Some(hrtime) = data.hrtime {
+        let s = hrtime / 1_000_000_000;
+        let ns = hrtime % 1_000_000_000;
+        println!("hrtime:    {s}.{ns:09}");
     }
 
-    println!("ipd_dataidx: {}", p.ipd_dataidx);
-    print!("ipd_data:    ");
-    if p.ipd_data.iter().all(|&c| c == 0) {
-        println!("[0; {}]", p.ipd_data.len());
-    } else {
-        match std::str::from_utf8(&p.ipd_data) {
-            Ok(s) => println!("{}", s.trim_matches('\0')),
-            Err(_e) => println!("{:?}", p.ipd_data),
+    println!("addr:      {:#x}", data.addr);
+    println!("pc:        {:#x}", data.pc);
+    println!("fp:        {:#x}", data.fp);
+    println!("rp:        {:#x}", data.rp);
+
+    if let Some(registers) = data.registers {
+        let registers = registers.iter().collect::<Vec<_>>();
+
+        println!("registers:");
+
+        for chunk in registers.chunks(3) {
+            print!("          ");
+
+            for (r, v) in chunk {
+                let label: String = format!("{r}").chars().take(3).collect();
+                print!(" {label:3} {v:#18x}");
+            }
+            println!();
         }
     }
 
+    println!("message:   {}", data.message.unwrap_or("<none>".to_string()));
```

Comment:
very tiny nit that doesn't actually matter in humility: since constructing the `unwrap_or` value is relatively expensive (allocates), maybe prefer lazily constructing it only if `data.message` actually is `None`?
```suggestion
    println!("message:   {}", data.message.unwrap_or_else(|| "<none>".to_string()));
```

Reply:
Fixed